gnovm/pkg/doc/doc.go

```diff
@@ -0,0 +1,303 @@
+// Package doc implements support for documentation of Gno packages and realms,
+// in a similar fashion to `go doc`.
+// As a reference, the [official implementation] for `go doc` is used.
+//
+// [official implementation]: https://github.com/golang/go/tree/90dde5dec1126ddf2236730ec57511ced56a512d/src/cmd/doc
+package doc
+
+import (
+	"fmt"
+	"go/ast"
+	"go/doc"
+	"go/token"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"go.uber.org/multierr"
+)
+
+// WriteDocumentationOptions represents the possible options when requesting
+// documentation through Documentable.
+type WriteDocumentationOptions struct {
+	// ShowAll shows all symbols when displaying documentation about a package.
+	ShowAll bool
+	// Source shows the source code when documenting a symbol.
+	Source bool
+	// Unexported shows unexported symbols as well as exported.
+	Unexported bool
+	// Short shows a one-line representation for each symbol.
+	Short bool
+
+	w io.Writer
+}
+
+// Documentable is a package, symbol, or accessible which can be documented.
+type Documentable interface {
+	WriteDocumentation(w io.Writer, opts *WriteDocumentationOptions) error
+}
+
+// static implementation check
+var _ Documentable = (*documentable)(nil)
+
+type documentable struct {
+	Dir
+	symbol     string
+	accessible string
+	pkgData    *pkgData
+}
+
+func (d *documentable) WriteDocumentation(w io.Writer, o *WriteDocumentationOptions) error {
+	if o == nil {
+		o = &WriteDocumentationOptions{}
+	}
+	o.w = w
+
+	var err error
+	// pkgData may already be initialised if we already had to look to see
+	// if it had the symbol we wanted; otherwise initialise it now.
+	if d.pkgData == nil {
+		d.pkgData, err = newPkgData(d.Dir, o.Unexported)
+		if err != nil {
+			return err
+		}
+	}
```

Comment:
Go doesn't actually allow this! (`expected identifier on the left side of :=`) When we're assigning to a struct field/indexed value we always need to use `=`...

Reply:
Right, my main problem is that `err` is outside the scope it is really used. Can we change it to at least move it inside the if context?:

```suggestion
	// pkgData may already be initialised if we already had to look to see
	// if it had the symbol we wanted; otherwise initialise it now.
	if d.pkgData == nil {
		var err error
		d.pkgData, err = newPkgData(d.Dir, o.Unexported)
		if err != nil {
			return err
		}
	}
```

Or even clearer IMHO:


```suggestion
	// pkgData may already be initialised if we already had to look to see
	// if it had the symbol we wanted; otherwise initialise it now.
	if d.pkgData == nil {
		pd, err := newPkgData(d.Dir, o.Unexported)
		if err != nil {
			return err
		}
		d.pkgData = pd
	}
```